funnel/utils/markdown/mdit_plugins/abbr.py

```diff
@@ -0,0 +1,173 @@
+"""
+Markdown-it-py plugin to introduce <abbr> markup for defined abbreviations.
+
+Ported from javascript plugin markdown-it-abbr.
+"""
+
+import re
+import unicodedata
+
+from markdown_it import MarkdownIt
+from markdown_it.rules_block import StateBlock
+from markdown_it.rules_inline import StateInline
+from markdown_it.token import Token
+
+__all__ = ['abbr_plugin']
+
+ASTERISK_CHAR = 0x2A  # ASCII value for `*`
+OPEN_SQBR_CHAR = 0x5B  # ASCII value for `[`
+CLOSE_SQBR_CHAR = 0x5D  # ASCII value for `]`
+BACKSLASH_CHAR = 0x5C  # ASCII value for `\`
+COLON_CHAR = 0x3A  # ASCII value for `:`
+
+PADDING_CHARACTERS_RE = (
+    '['
+    + ''.join(
+        chr(x) for x in range(65536) if unicodedata.category(chr(x))[0] in ('P', 'Z')
+    )
+    .replace('\\', '\\\\')
+    .replace('[', '\\[')
+    .replace(']', '\\]')
+    + '\r\n$+<=>^`|~'
+    + ']'
+)
+
+
+def abbr_def(state: StateBlock, start_line: int, end_line: int, silent: bool):
+    pos = state.bMarks[start_line] + state.tShift[start_line]
+    maximum = state.eMarks[start_line]
+
+    if pos + 2 >= maximum:
+        return False
+
+    # Check if line is of the format `*[Abbreviation]:Full Form`
+    if state.srcCharCode[pos] != ASTERISK_CHAR:
+        return False
+    pos += 1
+    if state.srcCharCode[pos] != OPEN_SQBR_CHAR:
+        return False
+    pos += 1
+
+    label_end = -1
+    label_start = pos
+    while pos < maximum:
+        ch = state.srcCharCode[pos]
+        if ch == OPEN_SQBR_CHAR:  # pylint: disable=no-else-return
+            return False
+        elif ch == CLOSE_SQBR_CHAR:
+            label_end = pos
+            break
+        elif ch == BACKSLASH_CHAR:
+            pos += 1
+        pos += 1
+
+    if label_end < 0 or state.srcCharCode[label_end + 1] != COLON_CHAR:
+        return False
+
+    if silent:
+        return True
+
+    # Extract label and title and store it in state.env
+    label = state.src[label_start:label_end].replace('\\', '')
+    title = state.src[label_end + 2 : maximum].strip()
+    if len(label) == 0 or len(title) == 0:
+        return False
+
+    if 'abbr' not in state.env:
+        state.env['abbr'] = {}
+
+    if label not in state.env['abbr']:
+        state.env['abbr'][label] = title
+
+    state.line = start_line + 1
+    return True
+
+
+def abbr_replace(state: StateInline):
+    block_tokens = state.tokens
+
+    if 'abbr' not in state.env:
+        return
+
+    labels_re_str = '|'.join(
+        [state.md.utils.escapeRE(k) for k in sorted(state.env['abbr'].keys(), key=len)]
+    )
+
+    simple_re = re.compile('(?:' + labels_re_str + ')')
+
+    match_re_str = (
+        '(^|' + PADDING_CHARACTERS_RE + ')'
+        '(' + labels_re_str + ')'
+        '($|' + PADDING_CHARACTERS_RE + ')'
+    )
```

Comment:
I just realised this padding char regex is unnecessary. This could simply be `(^|\W)` and `($|\W)`, with the regex compiled using `re.UNICODE` flag.

[Documentation on `\W`](https://docs.python.org/3/library/re.html#index-35).

Reply:
https://docs.python.org/3/library/re.html#re.ASCII
> Note that for backward compatibility, the re.U flag still exists (as well as its synonym re.UNICODE and its embedded counterpart (?u)), but these are redundant in Python 3 since matches are Unicode by default for strings (and Unicode matching isn’t allowed for bytes).